web-api/src/middleware/tracing.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+/// Like [`HttpServer.new()`] but sets things up to use a given non-global tracing subscriber.
+///
+/// The wrapper will do following things:
+///
+/// 1. wrap the call of the factory function passed to [`HttpServer.new()`] so that
+///    the right subscriber is active during that call.
+/// 2. add a middleware which
+///     a. provides the right subscriber to all middleware calls (e.g. a `warp_fn` function
+///         will have the right logging context even before calling `service.call(request)`)
+///     b. wraps the build request future so that it has the right logging context when resolving
+// Hint: Making this a macro saves us very complex nightmare of type annotations (and also way less lines of code).
```

Comment:
I did try, but the way actix-web does handle certain type system things wrt. `App`/`HttpServer` creation makes it just absurdly painful to wrap them. This also includes not exposing some pub type which is part of the type signature.

Reply:
i agree, this part of actix really isn't nice to be wrapped/reused with explicit type annotations.